File: 剑指offer/59 - II. 队列的最大值.py
"""
执行用时 :
284 ms
, 在所有 Python3 提交中击败了
62.24%
的用户
内存消耗 :
17.3 MB
, 在所有 Python3 提交中击败了
100.00%
的用户
"""

class MaxQueue:

    def __init__(self):
        self.queue = collections.deque()

    def max_value(self) -> int:
        return max(self.queue) if self.queue else -1

    def push_back(self, value: int) -> None:
        self.queue.append(value)

    def pop_front(self) -> int:
        return self.queue.popleft() if self.queue else -1

File: 剑指offer/C++_Version/59 - II. 队列的最大值.cpp
//方法1
class MaxQueue {
public:
    MaxQueue() {}
    deque<int> deq;

    int max_value() 
    {   int max_val=deq[0];
        for(int i=1; i<deq.size(); i++)
        {
            if(deq[i]>max_val)
                max_val = deq[i];
        }

        if(!deq.empty())
            return max_val;
        else
            return -1;
    }
    
    void push_back(int value) 
    {
        deq.push_back(value);
    }
    
    int pop_front()
    {
        if(!deq.empty())
        {
            int front = deq.front();
            deq.pop_front();
            return front;
        }
        else
            return -1;
    }
};

/**
 * Your MaxQueue object will be instantiated and called as such:
 * MaxQueue* obj = new MaxQueue();
 * int param_1 = obj->max_value();
 * obj->push_back(value);
 * int param_3 = obj->pop_front();
 */
